[socket-mode] webClient error is swallowed

### Description

When `webClient.apiCall` throws an error it is not propagated and handled by `onFailure`. Instead an error is logged and forgotten.

```ts
              return this.webClient.apiCall('apps.connections.open').then((result: WebAPICallResult) => {
                return result;
              }).catch((error) => {
                this.logger.error(error);
              });
```

As a result machine changes state to `authenticated` handled where `context.result` is `undefined`

Which ultimately leads to `TypeError: Cannot read property 'url' of undefined` in 

```ts
                this.setupWebSocket(context.result.url);
```

### What type of issue is this? (place an `x` in one of the `[ ]`)
- [x] bug
- [ ] enhancement (feature request)
- [ ] question
- [ ] documentation related
- [ ] testing related
- [ ] discussion

### Requirements (place an `x` in each of the `[ ]`)
* [x] I've read and understood the [Contributing guidelines](https://github.com/slackapi/node-slack-sdk/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

---

### Bug Report

Filling out the following details about bugs will help us solve your issue sooner.

#### Packages:

Select all that apply:

- [ ] `@slack/web-api`
- [ ] `@slack/events-api`
- [ ] `@slack/interactive-messages`
- [ ] `@slack/rtm-api`
- [ ] `@slack/webhooks`
- [ ] `@slack/oauth`
- [ ] I don't know
- [x] `@slack/socket-mode`

#### Reproducible in:

package version: 1.0.0

node version: v12.18.4

OS version(s): macOS Catalina

#### Steps to reproduce:

1. Provide wrong app token.
2. Start bolt with `socketMode: true`

#### Expected result:

Error is rethrown after logging and state machine does not transition to "authenticated".

#### Actual result:

Error is swallowed and state machine transitions to "authenticated".

#### Attachments:

```
⚡️ Bolt app started
[DEBUG]  web-api:WebClient:0 http response received
[ERROR]  socket-mode:SocketModeClient:0 Error: An API error occurred: not_authed
    at Object.platformErrorFromResult (/Users/#####/node_modules/@slack/socket-mode/node_modules/@slack/web-api/dist/errors.js:51:33)
    at WebClient.apiCall (/Users/#####/node_modules/@slack/socket-mode/node_modules/@slack/web-api/dist/WebClient.js:156:28)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  code: 'slack_webapi_platform_error',
  data: {
    ok: false,
    error: 'not_authed',
    response_metadata: { acceptedScopes: [Array] }
  }
}
[DEBUG]  socket-mode:SocketModeClient:0 transitioning to state: connecting:authenticated
(node:92065) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'url' of undefined
    at /Users/#####/node_modules/@slack/socket-mode/dist/SocketModeClient.js:76:48
    at /Users/#####/node_modules/finity/lib/utils/invokeEach.js:11:15
    at Array.forEach (<anonymous>)
    at invokeEach (/Users/#####/node_modules/finity/lib/utils/invokeEach.js:10:7)
    at StateMachine.enterState (/Users/#####/node_modules/finity/lib/core/StateMachine.js:125:32)
    at StateMachine.executeTransition (/Users/#####/node_modules/finity/lib/core/StateMachine.js:116:12)
    at /Users/#####/node_modules/finity/lib/core/StateMachine.js:285:16
    at TaskScheduler.execute (/Users/#####/node_modules/finity/lib/core/TaskScheduler.js:29:7)
    at StateMachine.executeTrigger (/Users/#####/node_modules/finity/lib/core/StateMachine.js:282:24)
    at StateMachine.handleAsyncActionComplete (/Users/#####/node_modules/finity/lib/core/StateMachine.js:195:10)
(node:92065) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:92065) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
[DEBUG]  web-api:WebClient:0 http response received
```

![](https://media.giphy.com/media/FEuvG7U6wNYEE/giphy.gif)

@stevengill 
I am not sure the following code is the best solution yet but I am thinking that we may want to have an error handling for `TypeError: Cannot read property 'url' of undefined` in this scenario. As the `apps.connections.open` API method call returns 200 OK even for an error response, the state machine goes to `.onSuccess().transitionTo('authenticated')` pattern. In this instance, the `context.result` can be unexpectedly undefined.

https://github.com/slackapi/node-slack-sdk/blob/%40slack/socket-mode%401.0.0/packages/socket-mode/src/SocketModeClient.ts#L99-L107

```js
.state('authenticated')
.onEnter((_state, context) => {
  this.authenticated = true;
  if (typeof context.result == "undefined") {
    // ok: false
    setImmediate(() => {
      this.emit('failed', context.result);
    });
  } else {
    this.setupWebSocket(context.result.url);
    setImmediate(() => {
      this.emit('authenticated', context.result);
    });
  }
})
.on('websocket open').transitionTo('handshaking')
```

With this change, `(async () => { await app.start(); })();` simply exists this way:

```
$ npx node index.js
⚡️ Bolt app started
[ERROR]  socket-mode:SocketModeClient:0 Error: An API error occurred: invalid_auth
    at Object.platformErrorFromResult (/Users/ksera/github/bolt-starter/node_modules/@slack/socket-mode/node_modules/@slack/web-api/dist/errors.js:51:33)
    at WebClient.apiCall (/Users/ksera/github/bolt-starter/node_modules/@slack/socket-mode/node_modules/@slack/web-api/dist/WebClient.js:156:28)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  code: 'slack_webapi_platform_error',
  data: { ok: false, error: 'invalid_auth', response_metadata: {} }
```

--- a/packages/socket-mode/package.json
+++ b/packages/socket-mode/package.json
@@ -44,8 +44,8 @@
     "watch": "npx nodemon --watch 'src' --ext 'ts' --exec npm run build"
   },
   "dependencies": {
-    "@slack/logger": "^2.0.0",
-    "@slack/web-api": "^5.14.0",
+    "@slack/logger": "^3.0.0",
+    "@slack/web-api": "^6.0.0",
     "@types/node": ">=12.0.0",
     "@types/p-queue": "^2.3.2",
     "@types/ws": "^7.2.5",

--- a/packages/socket-mode/src/SocketModeClient.ts
+++ b/packages/socket-mode/src/SocketModeClient.ts
@@ -67,22 +67,22 @@ export class SocketModeClient extends EventEmitter {
           .initialState('authenticating')
             .do(() => {
               // TODO: change this to webClient.apps.connections.open()
-              return this.webClient.apiCall('apps.connections.open').then((result: WebAPICallResult) => {
+              return this.webClient.apps.connections.open().then((result: WebAPICallResult) => {
                 return result;
               }).catch((error) => {
                 this.logger.error(error);
+                // throw error;
+                return Promise.reject(error);
               });
             })
               .onSuccess().transitionTo('authenticated')
               .onFailure()
                 .transitionTo('reconnecting').withCondition((context) => {
                   const error = context.error as WebAPICallError;
-                  this.logger.info('here');
                   this.logger.info(`unable to Socket Mode start: ${error.message}`);
 
                   // Observe this event when the error which causes reconnecting or disconnecting is meaningful
                   this.emit('unable_to_socket_mode_start', error);
-
                   let isRecoverable = true;
                   if (error.code === APICallErrorCode.PlatformError &&
                       (Object.values(UnrecoverableSocketModeStartError) as string[]).includes(error.data.error)) {
@@ -156,17 +156,18 @@ export class SocketModeClient extends EventEmitter {
             .initialState('authenticating')
               .do(() => {
                 // TODO: change this to webClient.apps.connections.open()
-                return this.webClient.apiCall('apps.connections.open').then((result: WebAPICallResult) => {
+                return this.webClient.apps.connections.open().then((result: WebAPICallResult) => {
                   return result;
                 }).catch((error) => {
                   this.logger.error(error);
+                  // throw error;
+                  return Promise.reject(error);
                 });
               })
                 .onSuccess().transitionTo('authenticated')
                 .onFailure()
                   .transitionTo('authenticating').withCondition((context) => {
                     const error = context.error as WebAPICallError;
-                    this.logger.info('here');
                     this.logger.info(`unable to Socket Mode start: ${error.message}`);
 
                     // Observe this event when the error which causes reconnecting or disconnecting is meaningful

--- a/packages/socket-mode/src/SocketModeClient.ts
+++ b/packages/socket-mode/src/SocketModeClient.ts
@@ -66,7 +66,6 @@ export class SocketModeClient extends EventEmitter {
         .submachine(Finity.configure()
           .initialState('authenticating')
             .do(() => {
-              // TODO: change this to webClient.apps.connections.open()
               return this.webClient.apps.connections.open().then((result: WebAPICallResult) => {
                 return result;
               }).catch((error) => {
@@ -155,7 +154,6 @@ export class SocketModeClient extends EventEmitter {
             .submachine(Finity.configure()
             .initialState('authenticating')
               .do(() => {
-                // TODO: change this to webClient.apps.connections.open()
                 return this.webClient.apps.connections.open().then((result: WebAPICallResult) => {
                   return result;
                 }).catch((error) => {

